README.md
# sec-cik-mapper

[![Tests](https://github.com/jadchaar/sec-cik-mapper/actions/workflows/continuous_integration.yml/badge.svg)](https://github.com/jadchaar/sec-cik-mapper/actions/workflows/continuous_integration.yml)
[![Update Mappings Daily CRON Job](https://github.com/jadchaar/cik-mapper/actions/workflows/update_mappings_daily_cron_job.yml/badge.svg?event=schedule)](https://github.com/jadchaar/sec-cik-mapper/actions/workflows/update_mappings_daily_cron_job.yml)
[![Documentation Status](https://readthedocs.org/projects/sec-cik-mapper/badge/?version=latest)](https://sec-cik-mapper.readthedocs.io/en/latest/?badge=latest)
[![Coverage](https://codecov.io/gh/jadchaar/sec-cik-mapper/branch/main/graph/badge.svg)](https://pypi.org/project/sec-cik-mapper/)

[![PyPI Version](https://img.shields.io/pypi/v/sec-cik-mapper.svg)](https://pypi.org/project/sec-cik-mapper/)
[![Supported Python Versions](https://img.shields.io/pypi/pyversions/sec-cik-mapper.svg)](https://pypi.org/project/sec-cik-mapper/)
[![License](https://img.shields.io/pypi/l/sec-cik-mapper.svg)](https://pypi.org/project/sec-cik-mapper/)
[![Code Style: Black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/python/black)

**sec-cik-mapper** is a Python package for generating mappings between stock and mutual fund identifier data provided by the SEC.

## Features

- Generate up-to-date mappings from the SEC as native Python dictionaries
- Mappings for both stocks and mutual funds
- Mapping data exposed as a raw pandas dataframe for custom data processing and usage
- Full support for PEP 484-style type hints and the [mypy type checker](https://mypy.readthedocs.io/en/stable/)
- [Pre-generated mappings](https://github.com/jadchaar/sec-cik-mapper/tree/main/mappings), updated daily, available from GitHub and jsDelivr for use outside of Python
- Support for Python 3.6+

## Quick Start

### Installation

Install and update this package using [pip](https://pip.pypa.io/en/stable/getting-started/):

```console
$ pip install -U sec-cik-mapper
```

### Basic Usage

#### Stocks

```python
>>> from sec_cik_mapper import StockMapper
>>> from pathlib import Path

# Initialize a stock mapper instance
>>> mapper = StockMapper()

# Get mapping from CIK to tickers
>>> mapper.cik_to_tickers
{'0000320193': {'AAPL'}, '0000789019': {'MSFT'}, '0001652044': {'GOOG', 'GOOGL'}, ...}

# Get mapping from ticker to CIK
>>> mapper.ticker_to_cik
{'AAPL': '0000320193', 'MSFT': '0000789019', 'GOOG': '0001652044', ...}

# Get mapping from CIK to company name
>>> mapper.cik_to_company_name
{'0000320193': 'Apple Inc.', '0000789019': 'Microsoft Corp', '0001652044': 'Alphabet Inc.', ...}

# Get mapping from ticker to company name
>>> mapper.ticker_to_company_name
{'AAPL': 'Apple Inc.', 'MSFT': 'Microsoft Corp', 'GOOG': 'Alphabet Inc.', ...}

# Get mapping from ticker to exchange
>>> mapper.ticker_to_exchange
{'AAPL': 'Nasdaq', 'MSFT': 'Nasdaq', 'GOOG': 'Nasdaq', ...}

# Get mapping from exchange to tickers
>>> mapper.exchange_to_tickers
{'Nasdaq': {'CYRN', 'OHPAW', 'SANW', ...}, 'NYSE': {'PLAG', 'TDW-WTB', 'RS', ...}, 'OTC': {'ZICX', 'LTGJ', 'AVNI', ...}, ...}

# Get mapping from CIK to exchange
>>> mapper.cik_to_exchange
{'0000320193': 'Nasdaq', '0000789019': 'Nasdaq', '0001652044': 'Nasdaq', ...}

# Get mapping from exchange to CIKs
>>> mapper.exchange_to_ciks
{'Nasdaq': {'0000779544', '0001508171', '0001060955', ...}, 'NYSE': {'0000764478', '0000008818', '0001725057', ...}, 'OTC': {'0001044676', '0001592411', '0001284452', ...}, ...}

# Save CIK, ticker, exchange, and company name mappings to a CSV file
>>> csv_path = Path("example_mappings.csv")
>>> mapper.save_metadata_to_csv(csv_path)

# Get raw pandas dataframe
>>> mapper.raw_dataframe
              CIK Ticker                               Name Exchange
0      0000001750    AIR                           Aar Corp     NYSE
1      0000001800    ABT                Abbott Laboratories     NYSE
2      0000001961   WDDD                         Worlds Inc      OTC
3      0000002098    ACU                   Acme United Corp     NYSE
4      0000002178     AE     Adams Resources & Energy, Inc.     NYSE
...           ...    ...                                ...      ...
13167  0001894057   PPYA  Papaya Growth Opportunity Corp. I
13168  0001894370   HMAC   Hainan Manaslu Acquisition Corp.
13169  0001894630   GHIX            Gores Holdings Ix, Inc.
13170  0001897941   GTEN             Gores Holdings X, Inc.
13171  0001898416   ALVO       Alvotech Lux Holdings S.A.S.

[13172 rows x 4 columns]
```

#### Mutual Funds

```python
>>> from sec_cik_mapper import MutualFundMapper
>>> from pathlib import Path

# Initialize a mutual fund mapper instance
>>> mapper = MutualFundMapper()

# Get mapping from CIK to tickers
>>> mapper.cik_to_tickers
{'0000002110': {'CRBYX', 'CEFZX', 'CSSRX', ...}, '0000002646': {'IIBPX', 'IPISX', 'IIBTX', ...}, '0000002663': {'IMSXX', 'VMTXX', 'IVMXX', ...}, ...}

# Get mapping from ticker to CIK
>>> mapper.ticker_to_cik
{'LACAX': '0000002110', 'LIACX': '0000002110', 'ACRNX': '0000002110', ...}

# Get mapping from CIK to series ID
>>> mapper.cik_to_series_ids
{'0000002110': {'S000009184', 'S000033622', 'S000009185', ...}, '0000002646': {'S000008760'}, '0000002663': {'S000008702'}, ...}

# Get mapping from ticker to series ID
>>> mapper.ticker_to_series_id
{'LACAX': 'S000009184', 'LIACX': 'S000009184', 'ACRNX': 'S000009184', ...}

# Get mapping from series ID to CIK
>>> mapper.series_id_to_cik
{'S000009184': '0000002110', 'S000009185': '0000002110', 'S000009186': '0000002110', ...}

# Get mapping from series ID to tickers
>>> mapper.series_id_to_tickers
{'S000009184': {'CEARX', 'CRBYX', 'ACRNX', ...}, 'S000009185': {'ACINX', 'CACRX', 'CAIRX', ...}, 'S000009186': {'LAUCX', 'LAUAX', 'CUSAX', ...}, ...}

# Get mapping from series ID to class IDs
>>> mapper.series_id_to_class_ids
{'S000009184': {'C000024956', 'C000122737', 'C000024957', ...}, 'S000009185': {'C000024958', 'C000122739', 'C000097733', ...}, 'S000009186': {'C000024962', 'C000024964', 'C000122740', ...}, ...}

# Get mapping from ticker to class ID
>>> mapper.ticker_to_class_id
{'LACAX': 'C000024954', 'LIACX': 'C000024956', 'ACRNX': 'C000024957', ...}

# Get mapping from CIK to class IDs
>>> mapper.cik_to_class_ids
{'0000002110': {'C000024958', 'C000024969', 'C000024957', ...}, '0000002646': {'C000023849', 'C000074893', 'C000028785', ...}, '0000002663': {'C000023718', 'C000028786', 'C000076529', ...}, ...}

# Get mapping from class ID to CIK
>>> mapper.class_id_to_cik
{'C000024954': '0000002110', 'C000024956': '0000002110', 'C000024957': '0000002110', ...}

# Get mapping from class ID to ticker
>>> mapper.class_id_to_ticker
{'C000024954': 'LACAX', 'C000024956': 'LIACX', 'C000024957': 'ACRNX', ...}

# Save CIK, ticker, series ID, and class ID mappings to a CSV file
>>> csv_path = Path("mutual_fund_mappings.csv")
>>> mapper.save_metadata_to_csv(csv_path)

# Get raw pandas dataframe
>>> mapper.raw_dataframe
              CIK Ticker   Series ID    Class ID
0      0000002110  ACFFX  S000009188  C000024973
1      0000002110  ACINX  S000009185  C000024961
2      0000002110  ACRNX  S000009184  C000024957
3      0000002110  ACTWX  S000009187  C000024969
4      0000002110  AUSAX  S000009186  C000024965
...           ...    ...         ...         ...
29236  0001860434   INNO  S000073580  C000230585
29237  0001860434   SIFI  S000072556  C000228889
29238  0001860434   SIHY  S000072555  C000228888
29239  0001877493    BTF  S000074058  C000231452
29240  0001877493    VBB  S000075054  C000233857

[29241 rows x 4 columns]
```

## Supported Mappings

Mappings can be formed between the following SEC identifiers and metadata:

|     Key     |       Value       | `StockMapper` | `MutualFundMapper` |
| :---------: | :---------------: | :-----------: | :----------------: |
|    `CIK`    |  `Set(Tickers)`   |       ✅       |         ✅          |
|    `CIK`    |  `Company Name`   |       ✅       |                    |
|    `CIK`    |    `Exchange`     |       ✅       |                    |
| `Exchange`  |    `Set(CIKs)`    |       ✅       |                    |
| `Exchange`  |  `Set(Tickers)`   |       ✅       |                    |
|  `Ticker`   |       `CIK`       |       ✅       |         ✅          |
|  `Ticker`   |  `Company Name`   |       ✅       |                    |
|  `Ticker`   |    `Exchange`     |       ✅       |                    |
|    `CIK`    | `Set(Series IDs)` |               |         ✅          |
|    `CIK`    | `Set(Class IDs)`  |               |         ✅          |
| `Class ID`  |       `CIK`       |               |         ✅          |
| `Class ID`  |     `Ticker`      |               |         ✅          |
| `Series ID` |       `CIK`       |               |         ✅          |
| `Series ID` | `Set(Class IDs)`  |               |         ✅          |
| `Series ID` |  `Set(Tickers)`   |               |         ✅          |
|  `Ticker`   |    `Class ID`     |               |         ✅          |
|  `Ticker`   |    `Series ID`    |               |         ✅          |

## Pre-generated Mappings

Pre-generated mappings are also available for download and use outside of Python (e.g. manually or via automated scripts/curl requests). These mappings are updated daily via a [fully automated daily CRON job](https://github.com/jadchaar/sec-cik-mapper/actions/workflows/update_mappings_daily_cron_job.yml), which fetches, transforms, validates, and uploads the latest mapping data from the SEC to GitHub (save location: [sec-cik-mapper/mappings · GitHub](https://github.com/jadchaar/sec-cik-mapper/tree/main/mappings)). These mappings are available for download and usage from both GitHub and the [jsDelivr CDN](https://www.jsdelivr.com).

### Example Usage

Example [curl](https://curl.se/) commands, which download the specified mapping files and saves them to the current working directory:

#### Stocks

Hosted via GitHub:

```console
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/stocks/mappings.csv -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/stocks/cik_to_exchange.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/stocks/cik_to_tickers.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/stocks/ticker_to_exchange.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/stocks/cik_to_company_name.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/stocks/ticker_to_cik.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/stocks/ticker_to_company_name.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/stocks/exchange_to_tickers.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/stocks/exchange_to_ciks.json -O
```

Hosted via jsDelivr CDN:

```console
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/stocks/mappings.csv -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/stocks/cik_to_exchange.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/stocks/cik_to_tickers.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/stocks/ticker_to_exchange.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/stocks/cik_to_company_name.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/stocks/ticker_to_cik.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/stocks/ticker_to_company_name.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/stocks/exchange_to_tickers.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/stocks/exchange_to_ciks.json -O
```

#### Mutual Funds

Hosted via GitHub:

```console
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/ticker_to_class_id.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/series_id_to_class_ids.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/mappings.csv -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/cik_to_class_ids.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/cik_to_series_ids.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/series_id_to_cik.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/ticker_to_series_id.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/cik_to_tickers.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/class_id_to_cik.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/series_id_to_tickers.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/class_id_to_ticker.json -O
$ curl https://raw.githubusercontent.com/jadchaar/sec-cik-mapper/main/mappings/mutual_funds/ticker_to_cik.json -O
```

Hosted via jsDelivr CDN:

```console
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/ticker_to_class_id.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/series_id_to_class_ids.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/mappings.csv -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/cik_to_class_ids.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/cik_to_series_ids.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/series_id_to_cik.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/ticker_to_series_id.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/cik_to_tickers.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/class_id_to_cik.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/series_id_to_tickers.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/class_id_to_ticker.json -O
$ curl https://cdn.jsdelivr.net/gh/jadchaar/sec-cik-mapper@main/mappings/mutual_funds/ticker_to_cik.json -O
```

### Contributing

If you encounter a bug or would like to see a new company filing or feature added to **sec-cik-mapper**, please [file an issue](https://github.com/jadchaar/sec-cik-mapper/issues) or [submit a pull request](https://help.github.com/en/articles/creating-a-pull-request).

### Documentation

For full documentation, please visit [sec-cik-mapper.readthedocs.io](https://sec-cik-mapper.readthedocs.io).


CHANGELOG.md
# Changelog

## 2.1.0 - 1/9/22

### New

- Added an LRU cache to all StockMapper and MutualFundMapper properties. This should result in performance improvements on repeated calls to the mapper properties.

## 2.0.2 - 12/29/21

### Internal

- PyPI releases are now fully automated via a GitHub actions workflow that is triggered on every new git tag.
- Cleaned up Makefile, requirements.txt file and tox.ini configurations for faster and cleaner builds.
- Broke up the `requirements.txt` file into 3 different files under the `requirements/` folder to separate runtime, testing, and documentation dependencies.

## 2.0.1 - 12/28/21

### Changed

- Moved README content from `README.rst` to `README.md` for improved table formatting support.
- Raw pandas dataframe is now sorted on CIK and ticker to produce more consistent output.

## 2.0.0 - 12/25/21

### New

- Full revamp of API. Renamed `CIKMapper` to `StockMapper` and added support for mutual fund mapping capabilities via the `MutualFundMapper`. The `MutualFundMapper` is capable of obtaining mappings between CIKs, tickers, series IDs, and class IDs. The `StockMapper` is capable of obtaining mappings between CIKs, tickers, company names, and exchanges.
- Replaced function calls on mapper objects with properties.
- Added support for creating mappings to and from exchanges via the `StockMapper`.
- Added full support for PEP 484-style type hints.
- Added `raw_dataframe` property to both mappers that exposes the underlying pandas dataframe for additional extensibility.
- Added [examples to GitHub repository](https://github.com/jadchaar/sec-cik-mapper/tree/main/examples) to improve first-time usage.
- Updated naming of JSON and CSV mappings and added support for a number of new auto-generated mappings.
- Improved documentation in README on how to download the auto-generated mappings via curl from GitHub or jsDelivr.

### Internal

- Improved abstractions to reduce code duplication across both mapper classes.
- Migrated PyPI publishing system to [Flit](https://flit.readthedocs.io/).
- Revamped README and documentation. The latest documentation can be found [here](https://sec-cik-mapper.readthedocs.io).
- Full rewrite of tests to support new properties and package APIs.

## 1.0.0 - 12/15/21

- Initial release.


sec_cik_mapper/BaseMapper.py
"""Provides a :class:`BaseMapper` class for mapping stock and mutual
fund data from the SEC."""

import time
from collections import defaultdict
from pathlib import Path
from typing import ClassVar, Dict, List, Union, cast

import pandas as pd
import requests

from .retrievers import MutualFundRetriever, StockRetriever
from .types import CompanyData, FieldIndices, Fields, KeyToValueSet
from .utils import with_cache


class BaseMapper:
    """A :class:`BaseMapper` object."""

    _headers: ClassVar[Dict[str, str]] = {
        "User-Agent": f"{int(time.time())} {int(time.time())}@gmail.com",
        "Accept-Encoding": "gzip, deflate",
        "Host": "www.sec.gov",
    }

    def __init__(self, retriever: Union[StockRetriever, MutualFundRetriever]) -> None:
        """Constructor for the :class:`BaseMapper` class."""
        self.retriever = retriever
        self.mapping_metadata = self._get_mapping_metadata_from_sec()

    def __new__(cls, *args, **kwargs):
        """BaseMapper should not be directly instantiated,
        so throw an error on instantiation.

        More info: https://stackoverflow.com/a/7990308/3820660
        """
        if cls is BaseMapper:
            raise TypeError(
                f"{cls.__name__} cannot be directly instantiated. "
                "Please instantiate the StockMapper and/or MutualFundMapper "
                "classes instead."
            )
        return object.__new__(cls, *args, **kwargs)

    def _get_indices_from_fields(self, fields: Fields) -> FieldIndices:
        """Get list indices from field names."""
        field_indices = {field: fields.index(field) for field in fields}
        return cast(FieldIndices, field_indices)

    def _get_mapping_metadata_from_sec(self) -> pd.DataFrame:
        """Get company mapping metadata from the SEC as a pandas dataframe,
        sorted by CIK and ticker.
        """
        resp = requests.get(self.retriever.source_url, headers=BaseMapper._headers)
        resp.raise_for_status()
        data = resp.json()

        fields: Fields = data["fields"]
        field_indices: FieldIndices = self._get_indices_from_fields(fields)

        company_data: CompanyData = data["data"]
        transformed_data: List[Dict[str, str]] = []

        for cd in company_data:
            transformed_data.append(
                self.retriever.transform(field_indices, cd),
            )

        df = pd.DataFrame(transformed_data)
        df.sort_values(by=["CIK", "Ticker"], inplace=True, ignore_index=True)
        return df

    def _form_kv_set_mapping(self, keys: pd.Series, values: pd.Series) -> KeyToValueSet:
        """Form mapping from key to list of values, ignoring blank keys and values.

        Example: numerous CIKs map to multiple tickers (e.g. Banco Santander),
        so we must keep a list of tickers for each unique CIK.
        """
        mapping = defaultdict(set)
        for key, value in zip(keys, values):
            # Ignore blank keys and values
            if key and value:
                mapping[key].add(value)
        return dict(mapping)

    def _form_kv_mapping(self, keys: pd.Series, values: pd.Series) -> Dict[str, str]:
        """Form key-value mapping, ignoring blank keys and values."""
        return {k: v for k, v in zip(keys, values) if k and v}

    @property  # type: ignore
    @with_cache
    def cik_to_tickers(self) -> KeyToValueSet:
        """Get CIK to tickers mapping.

        Usage::

            >>> from sec_cik_mapper import MutualFundMapper, StockMapper
            >>> stock_mapper = StockMapper()
            >>> mutual_fund_mapper = MutualFundMapper()
            >>> stock_mapper.cik_to_tickers
            {'0000320193': {'AAPL'}, '0001652044': {'GOOG', 'GOOGL'}, ...}
            >>> mutual_fund_mapper.cik_to_tickers
            {'0000002110': {'CRBYX', 'CEFZX', ...}, '0000002646': {'IIBPX', 'IPISX', ...}, ...}
        """
        cik_col = self.mapping_metadata["CIK"]
        ticker_col = self.mapping_metadata["Ticker"]
        return self._form_kv_set_mapping(cik_col, ticker_col)

    @property  # type: ignore
    @with_cache
    def ticker_to_cik(self) -> Dict[str, str]:
        """Get ticker to CIK mapping.

        Usage::

            >>> from sec_cik_mapper import MutualFundMapper, StockMapper
            >>> stock_mapper = StockMapper()
            >>> mutual_fund_mapper = MutualFundMapper()
            >>> stock_mapper.ticker_to_cik
            {'AAPL': '0000320193', 'MSFT': '0000789019', 'GOOG': '0001652044', ...}
            >>> mutual_fund_mapper.ticker_to_cik
            {'LACAX': '0000002110', 'LIACX': '0000002110', 'ACRNX': '0000002110', ...}
        """
        cik_col = self.mapping_metadata["CIK"]
        ticker_col = self.mapping_metadata["Ticker"]
        return self._form_kv_mapping(ticker_col, cik_col)

    @property  # type: ignore
    def raw_dataframe(self) -> pd.DataFrame:
        """Get raw pandas dataframe.

        Usage::

            >>> from sec_cik_mapper import MutualFundMapper, StockMapper
            >>> stock_mapper = StockMapper()
            >>> mutual_fund_mapper = MutualFundMapper()
            >>> stock_mapper.raw_dataframe
                        CIK  Ticker                                  Name Exchange
            0      0000320193    AAPL                            Apple Inc.   Nasdaq
            1      0000789019    MSFT                        Microsoft Corp   Nasdaq
            2      0001652044    GOOG                         Alphabet Inc.   Nasdaq
            3      0001018724    AMZN                        Amazon Com Inc   Nasdaq
            4      0001318605    TSLA                           Tesla, Inc.   Nasdaq
            ...           ...     ...                                   ...      ...
            13184  0001866816   OLITU             Omnilit Acquisition Corp.   Nasdaq
            13185  0001870778   OHAAU               Opy Acquisition Corp. I   Nasdaq
            13186  0001873324   PEPLW    Pepperlime Health Acquisition Corp   Nasdaq
            13187  0001877557  WEL-UN  Integrated Wellness Acquisition Corp     NYSE
            13188  0001877787  ZGN-WT   Ermenegildo Zegna Holditalia S.P.A.     NYSE

            [13189 rows x 4 columns]
            >>> mutual_fund_mapper.raw_dataframe
                            CIK Ticker   Series ID    Class ID
            0      0000002110  LACAX  S000009184  C000024954
            1      0000002110  LIACX  S000009184  C000024956
            2      0000002110  ACRNX  S000009184  C000024957
            3      0000002110  CEARX  S000009184  C000122735
            4      0000002110  CRBRX  S000009184  C000122736
            ...           ...    ...         ...         ...
            29237  0001860434   SIHY  S000072555  C000228888
            29238  0001860434   SIFI  S000072556  C000228889
            29239  0001860434   INNO  S000073580  C000230585
            29240  0001877493    BTF  S000074058  C000231452
            29241  0001877493    VBB  S000075054  C000233857

            [29242 rows x 4 columns]
        """
        return self.mapping_metadata

    def save_metadata_to_csv(self, path: Union[str, Path]) -> None:
        """Save stock mapping metadata (CIK, ticker, exchange, and company name)
        or mutual fund mapping metadata (CIK, ticker, series ID, class ID) from
        SEC to CSV.

        Usage::

            >>> from sec_cik_mapper import StockMapper, MutualFundMapper
            >>> from pathlib import Path
            >>> stock_mapper = StockMapper()
            >>> mutual_fund_mapper = MutualFundMapper()
            >>> csv_path = Path("cik_mapping.csv")
            # Save full CIK, ticker, exchange, and company name mapping to a CSV file
            >>> stock_mapper.save_metadata_to_csv(csv_path)
            # Save full CIK, ticker, series ID, and class ID mapping to a CSV file
            >>> mutual_fund_mapper.save_metadata_to_csv(csv_path)
        """
        self.mapping_metadata.to_csv(path, index=False)
